feat(balance): Sound handling overhaul

Retagin opened the pull request:
DRAFT PR DO NOT MERGE UNTIL UPDATED TO BE COMPATIBLE WITH MAIN


## Purpose of change (The Why)

Game handles sound in a very calc intensive way if more than a few monsters or NPCs are inside the reality bubble.
Sounds are handled in tiles, and the way that the volume of explosions or high damage/range gun sounds are calced means that a .30-07 anywhere on the current z-level or a landmine going off anywhere within the reality bubble will automatically deafen any character.

Sound volume is very inconsistent, with several very different philosophies as to volume magnitude. Sometimes volume is sorta equivalent to real life dB, sometimes it is just a vauge tile amount. A huge explosion is counted as sound 15 and above, generic landmines put out 6000 sound, flashbangs have 14 sound, and hulks walking is something like 15? sound.

Sounds decay linearly with distance, and there is a significant increase in the number of calcs required for each sound added to inform monsters/npcs.

NPCs react to sounds every time the player takes an action, which includes making LOS checks to the player. 
Monsters have code to let them attempt to follow sounds for more than one turn, but this is not effectively used. Monster wander_to positions are re-written potentially 10 times a turn, and whichever random sound was checked last is used.

## Describe the solution (The How)

Complete overhaul of sound system to use consistent measurement (dB spl)

terrain sound absorbtion value put into map::level_cache, along with function to zero and generate it on game start and terrain update.
Sounds are "flood filled" to build out sound_caches, game stores a vector of these. Sound volume decreases with distance roughly following the inverse square law, with additional losses from terrain sound absorbtion as applicable. 


Monsters, NPCs, and the player avatar all reference this vector to pull the tile specific volume of various sounds.
Monsters get fed the loudest valid sound in their tile, and check to see if it is more interesting than their current wander_to sound.
NPCs each get fed a filtered set of sounds, only react to sounds once per turn, and have a few optimizations in their code to hopefully improve performance.
The PC gets sounds as per normal. 

sound_caches are culled twice per turn, and are valid for culling if they have been heard both by monsters and by the player. This is tracked with a pair of bools per cache. 

Sound values for hard-coded functions updated to new system.

## Describe alternatives you've considered

Just capping sound at something reasonable and getting my sleepless nights back.
Doing an equivalent to a lightcast but using the terrain absorption values with a straight line cast.
Using an adaptation of the old system moved to polar coordinates, with some math shenanagins to improve performance.

## Testing

Updating to main, correcting bugs, then testing.

## Additional context

This was all because I just wanted gun volume to not be wackadoodle and for silencers + subsonic ammo to actually do something.

## Checklist

<!--
NOTE: Please grant permission for repository maintainers to edit your PR.  It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them.  In web UI, you can do it by clicking the "Allow edits and access to secrets by maintainers" checkbox next to "Create Pull Request" button at the bottom of the editor, or by clicking the same checkbox in the sidebar after PR has been created.

NOTE: Please read your emails. Anyone mentioned on Github with an @ will receive an email, any activity on your work will also send emails. This is more reliable than being notified on our Discord, you will always get an email.
--->

### Mandatory

- [X] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/contribute/changelog_guidelines/).
- [X] I ran the [code formatter](https://docs.cataclysmbn.org/contribute/contributing/#code-style).
- [ ] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/contribute/contributing/#pull-request-notes) like `closes #1234` in [Summary of the PR](#purpose-of-change-the-why) so it can be closed automatically.
- [X] I've [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

<!--
please remove sections irrelevant to this PR.

### Optional

- [ ] This PR ports commits from DDA or other cataclysm forks.
  - [ ] I have attributed original authors in the commit messages adding [`Co-Authored-By`](https://docs.github.com/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) in the commit message.
  - [ ] I have linked the URL of original PR(s) in the description.
- [ ] This is a C++ PR that modifies JSON loading or behavior.
  - [ ] I have documented the changes in the appropriate location in the `doc/` folder.
  - [ ] If documentation for this feature does not exist, please write it or at least note its lack in PR description.
  - [ ] New localizable fields need to be added to the `lang/bn_extract_json_strings.sh` script if it does not support them yet.
  - [ ] If applicable, add checks on game load that would validate the loaded data.
  - [ ] If it modifies format of save files, please add migration from the old format.
- [ ] This is a PR that modifies build process or code organization.
  - [ ] Please document the changes in the appropriate location in the `doc/` folder.
  - [ ] If documentation for this feature or process does not exist, please write it.
  - [ ] If the change alters versions of software required to build or work with the game, please document it.
- [ ] This is a PR that removes JSON entities.
  - [ ] The removed JSON entities have new entries in `data/json/obsoletion/` folder or use some other migration process for existing saves.
- [ ] This PR modifies BN's lua API.
  - [ ] I have committed the output of `deno task docs:gen` so the Lua API documentation is updated.
-->

Retagin committed:
Updated files with new sound code

Does not include JSON updates/changes.

scarf005 commented on src/sounds.h:
```diff
@@ -169,5 +193,100 @@ template<>
+// With this, we should be able to do everything with addition/subtraction.
+// Distance 1 only happens at the source of a sound, i.e., the reference volume.
+//
+// Is this horrific? Yes.
+// Its abjectly heinous, and an afront in the eyes of any diety that would stoop itself to such a point as to cast its gaze upon this wretched datum.
+// Do I care? Yes, it brings me great shame.
+const std::array<short, 122> dist_vol_loss = { 0, 1500, 602, 352, 250, 194, 158, 134, 116, 102, 92, 83, 76, 70, 64, 60, 56, 53, 50, 47, 45, 42, 40, 39, 37, 35, 34, 33, 32, 30, 29, 28, 28, 27, 26, 25, 24, 24, 23, 23, 22, 21, 21, 20, 20, 20, 19, 19, 18, 18, 18, 17, 17, 17, 16, 16, 16, 15, 15, 15, 15, 14, 14, 14, 14, 13, 13, 13, 13, 13, 12, 12, 12, 12, 12, 12, 12, 11, 11, 11, 11, 11, 11, 11, 10, 10, ...
```
```suggestion
constexpr auto dist_vol_loss = std::array<short, 122>{ 0, 1500, 602, 352, 250, 194, 158, 134, 116, 102, 92, 83, 76, 70, 64, 60, 56, 53, 50, 47, 45, 42, 40, 39, 37, 35, 34, 33, 32, 30, 29, 28, 28, 27, 26, 25, 24, 24, 23, 23, 22, 21, 21, 20, 20, 20, 19, 19, 18, 18, 18, 17, 17, 17, 16, 16, 16, 15, 15, 15, 15, 14, 14, 14, 14, 13, 13, 13, 13, 13, 12, 12, 12, 12, 12, 12, 12, 11, 11, 11, 11, 11, 11, 11, 10, 10, 10, 10, 10, 10, 10, 10, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 7, 7, 7, 7 };
```
[nitpick]
1. pre-caclculating known amount of values for performance win is a good thing.
2. `constexpr`

scarf005 commented on src/sounds.h:
```diff
@@ -35,52 +40,71 @@ enum class sound_t : int {
-            const std::string &variant = "default" );
-/** Functions identical to sound(..., true). */
-void ambient_sound( const tripoint &p, int vol, sound_t category, const std::string &description );
+void sound( const tripoint &p, short vol, sound_t category, const std::string &description,
+            bool movement_noise = false, bool from_player = false, bool from_monster = false,
+            bool from_npc = false, const std::string &id = "",
+            const std::string &variant = "default", const faction_id faction = faction_id( "no_faction" ),
+            const mfaction_str_id monfaction = mfaction_str_id( "" ) );
+void sound( const tripoint &p, short vol, sound_t category, const translation &description,
+            bool movement_noise = false, bool from_player = false, bool from_monster = false,
+            bool from_npc = false, const std::string &id = "",
+            const std::string &variant = "default", const faction_id faction = faction_id( "no_faction" ),
+            const mfaction_str_id monfaction = mfaction_str_id( "" ) );
```
this function is getting really large, i'd suggest making it just accept existing `sound_event` struct:
```c++
sound::sound({ .vol = 100, /* ... */  }); 
```

Retagin committed:
Updating code to main compatability, worked in comments

autofix-ci[bot] committed:
style(autofix.ci): automated formatting

autofix-ci[bot] committed:
style(autofix.ci): automated formatting (attempt 2/3)

autofix-ci[bot] committed:
style(autofix.ci): automated formatting (attempt 3/3)

scarf005 commented on src/version.h:
i thought it was `.gitignore`d? anyways it musn't be committed

Retagin committed:
Delete src/version.h

Should have been gitignored, but autofix screams into the void.

Retagin committed:
Update to sound function

Updates sounds::sound to take a sound_event
Merged all sound methods to use this
A few minor bug fixes

Retagin committed:
Fixed reload noise volume loading issue

Retagin committed:
updated speech and field volume values

Retagin committed:
fix: Fixed ter/furn id check in map

Retagin committed:
Merge branch 'main' into Sound-overhaul-testing

autofix-ci[bot] committed:
style(autofix.ci): automated formatting
